File: lib/models/ingredient_draft.dart
import '../providers/inventory_provider.dart' show expiryLabelFor;
import '../utils/expiry_calculator.dart';
import 'draft_field.dart';
import 'ingredient.dart';
import 'storage_area.dart';

class IngredientDraft {
  IngredientDraft({
    required this.id,
    required this.name,
    required this.quantity,
    required this.unit,
    required this.category,
    required this.storage,
    required this.shelfLifeDays,
    this.selected = true,
  });

  final String id;
  DraftField<String> name;
  DraftField<String> quantity;
  DraftField<String> unit;
  DraftField<String?> category;
  DraftField<IconType?> storage;
  DraftField<int?> shelfLifeDays;
  bool selected;

  Ingredient toIngredient() {
    final days = shelfLifeDays.value;
    final today = DateTime.now();
    final expiry = days == null ? null : today.add(Duration(days: days));
    final freshness = expiry == null
        ? 0.85
        : expiryFreshness(expiryDate: expiry, totalShelfLifeDays: days ?? 7);
    return Ingredient(
      name: name.value,
      quantity: quantity.value,
      unit: unit.value,
      imageUrl: '',
      freshnessPercent: freshness,
      state: freshnessStateForExpiry(freshness: freshness, expiryDate: expiry),
      category: category.value,
      storage: storage.value ?? IconType.fridge,
      expiryDate: expiry,
      expiryLabel: expiry == null ? '新鲜' : expiryLabelFor(expiry),
      shelfLifeDays: days,
    );
  }
}

File: apps/ios/FreshPantry/Domain/Models/IngredientDraft.swift
import Foundation

/// AI/manual intake draft for one ingredient before confirmation; converts to
/// an `Ingredient`. Transient (no equality/JSON).
struct IngredientDraft {
    var id: String
    var name: DraftField<String>
    var quantity: DraftField<String>
    var unit: DraftField<String>
    var category: DraftField<String?>
    var storage: DraftField<IconType?>
    var shelfLifeDays: DraftField<Int?>
    var selected: Bool

    init(
        id: String,
        name: DraftField<String>,
        quantity: DraftField<String>,
        unit: DraftField<String>,
        category: DraftField<String?>,
        storage: DraftField<IconType?>,
        shelfLifeDays: DraftField<Int?>,
        selected: Bool = true
    ) {
        self.id = id
        self.name = name
        self.quantity = quantity
        self.unit = unit
        self.category = category
        self.storage = storage
        self.shelfLifeDays = shelfLifeDays
        self.selected = selected
    }

    /// Mirrors Dart `toIngredient()`: derives expiry/freshness from shelfLifeDays.
    func toIngredient(now: Date = Date()) -> Ingredient {
        let days = shelfLifeDays.value
        let expiry = days.map { now.addingTimeInterval(TimeInterval($0 * 86400)) }
        let freshness: Double
        if let expiry {
            freshness = ExpiryCalculator.expiryFreshness(
                expiryDate: expiry,
                totalShelfLifeDays: days ?? 7,
                now: now
            )
        } else {
            freshness = 0.85
        }
        return Ingredient(
            name: name.value,
            quantity: quantity.value,
            unit: unit.value,
            imageUrl: "",
            freshnessPercent: freshness,
            state: ExpiryCalculator.freshnessStateForExpiry(
                freshness: freshness,
                expiryDate: expiry,
                now: now
            ),
            expiryLabel: expiry == nil ? String(localized: "expiry.fresh") : ExpiryCalculator.expiryLabelFor(expiry!, now: now),
            category: category.value,
            storage: storage.value ?? .fridge,
            expiryDate: expiry,
            shelfLifeDays: days
        )
    }
}
